Lib/ufo2ft/featureWriters/kernFeatureWriter.py

```diff
@@ -561,3 +521,81 @@ def partitionByScript(
             localSide2,
             pair.value,
         )
+
+
+def makeAllGlyphClassDefinitions(kerningPerScript, context, feaFile=None):
+    side1Classes = {}
+    side2Classes = {}
+    side1Membership = context.side1Membership
+    side2Membership = context.side2Membership
+
+    classDefs = {}
+    if feaFile is not None:
+        classNames = {cdef.name for cdef in ast.iterClassDefinitions(feaFile)}
+    else:
+        classNames = set()
+
+    # Generate common class names first so that common classes are correctly
+    # named in other lookups.
+    if COMMON_SCRIPT in kerningPerScript:
+        common_pairs = kerningPerScript[COMMON_SCRIPT]
+        for pair in common_pairs:
+            if pair.firstIsClass and pair.side1 not in side1Classes:
+                addClassDefinition(
+                    "kern1",
+                    pair.side1,
+                    side1Classes,
+                    side1Membership,
+                    classDefs,
+                    classNames,
+                    "Common",
+                )
+            if pair.secondIsClass and pair.side2 not in side2Classes:
+                addClassDefinition(
+                    "kern2",
+                    pair.side2,
+                    side2Classes,
+                    side2Membership,
+                    classDefs,
+                    classNames,
+                    "Common",
+                )
+
+    for script, pairs in kerningPerScript.items():
+        if script == COMMON_SCRIPT:
+            continue
+        for pair in pairs:
+            if pair.firstIsClass and pair.side1 not in side1Classes:
+                addClassDefinition(
+                    "kern1",
+                    pair.side1,
+                    side1Classes,
+                    side1Membership,
+                    classDefs,
+                    classNames,
+                    script,
+                )
+            if pair.secondIsClass and pair.side2 not in side2Classes:
+                addClassDefinition(
+                    "kern2",
+                    pair.side2,
+                    side2Classes,
+                    side2Membership,
+                    classDefs,
+                    classNames,
+                    script,
+                )
+
+    return classDefs, side1Classes, side2Classes
```

Comment:
`classDefs` isn't this simply the union of `side1Classes` and `side2Classes`, or am I missing something else?

Reply:
oh no sorry, one is keyed on the className the other on the group content itself (to dedup)

Reply:
It would have been obvious had I been using typing.